cryptol-remote-api/python/cryptol/quoting.py

```diff
@@ -0,0 +1,84 @@
+"""Cryptol quasiquotation using an f-string-like syntax"""
+
+from typing import Any, Union
+
+from .bitvector import BV
+from .opaque import OpaqueValue
+from .commands import CryptolValue
+from .cryptoltypes import CryptolCode, CryptolLiteral, parenthesize
+from .custom_fstring import *
+
+def to_cryptol_str(val : Union[CryptolValue, str, CryptolCode]) -> str:
+    """Converts a ``CryptolValue``, string literal, or ``CryptolCode`` into
+       a string of cryptol syntax."""
+    if isinstance(val, bool):
+        return 'True' if val else 'False'
+    elif isinstance(val, tuple):
+        return '(' + ', '.join(to_cryptol_str(x) for x in val) + ')'
+    elif isinstance(val, dict):
+        return '{' + ', '.join(f'{k} = {to_cryptol_str(v)}' for k,v in val.items()) + '}'
+    elif isinstance(val, int):
+        return str(val)
+    elif isinstance(val, list):
+        return '[' + ', '.join(to_cryptol_str(x) for x in val) + ']'
+    elif isinstance(val, BV):
+        if val.size() % 4 == 0:
+            return val.hex()
+        else:
+            return f'({val.to_signed_int()} : [{val.size()}])'
+    elif isinstance(val, OpaqueValue):
+        return str(val)
+    elif isinstance(val, str):
+        return f'"{val}"'
+    elif isinstance(val, CryptolCode):
+        return parenthesize(str(val))
+    else:
+        raise TypeError("Unable to convert value to Cryptol syntax: " + str(val))
+
+def to_cryptol_str_customf(s : str, *, frames : int = 0,
+                                       filename : str = "<cry_f>") -> str:
+    """The function used to parse strings given to ``cry_f``"""
+    return func_customf(s, to_cryptol_str, frames=1+frames,
+                                           filename=filename)
+
+def cry(s : str) -> CryptolLiteral:
+    """Embed a string of cryptol syntax as ``CryptolCode``"""
+    return CryptolLiteral(s)
+
+def cry_f(s : str) -> CryptolLiteral:
+    """Embed a string of cryptol syntax as ``CryptolCode``, where the given
+       string is parsed as an f-string, and the values within brackets are
+       converted to cryptol syntax using ``to_cryptol_str``.
```

Comment:
This docstring is a bit inconsistent about whether it capitalizes the C in Cryptol or not in English prose. I'd prefer using capital-C Cryptol whenever it's being used in a context where snake case isn't being used (i.e., not in a Python function name). That is,

```suggestion
    """Embed a string of Cryptol syntax as ``CryptolCode``, where the given
       string is parsed as an f-string, and the values within brackets are
       converted to Cryptol syntax using ``to_cryptol_str``.
```

And similarly for the rest of the docstring.